libsolidity/lsp/Transport.h

```diff
@@ -0,0 +1,101 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+#pragma once
+
+#include <json/value.h>
+
+#include <functional>
+#include <iosfwd>
+#include <map>
+#include <optional>
+#include <string>
+#include <string_view>
+#include <variant>
+
+namespace solidity::lsp
+{
+
+using MessageID = Json::Value;
+
+enum class ErrorCode
+{
+	// Defined by JSON RPC
+	ParseError = -32700,
+	MethodNotFound = -32601,
+	InvalidParams = -32602,
+	InternalError = -32603,
+
+	// Defined by the protocol.
+	ServerNotInitialized = -32002,
+	RequestFailed = -32803
+};
+
+/**
+ * Transport layer API
+ *
+ * The transport layer API is abstracted to make LSP more testable as well as
+ * this way it could be possible to support other transports (HTTP for example) easily.
+ */
+class Transport
+{
+public:
+	virtual ~Transport() = default;
+
+	virtual bool closed() const noexcept = 0;
+	virtual std::optional<Json::Value> receive() = 0;
+	virtual void notify(std::string _method, Json::Value _params) = 0;
+	virtual void reply(MessageID _id, Json::Value _result) = 0;
+	virtual void error(MessageID _id, ErrorCode _code, std::string _message) = 0;
+};
+
+/**
+ * LSP Transport using JSON-RPC over iostreams.
+ */
+class JSONTransport: public Transport
```

Comment:
Looks like https://github.com/ethereum/solidity/pull/11350#discussion_r762247646 has not been addressed.

I really think that this being named `JSONTransport` is confusing. Both it and `Transport` use JSON. The difference is really that the former uses iostreams and is a concrete implementation while the latter is an interface. I think that the names do not reflect that at all.

Reply:
JSON objects, I agree, and this one is about is a JSON (RPC) transport. So `Transport` itself is just the interface. So the difference is, that this class is actually implementing this interface with respect to JSON-RPC. :)

Reply:
Isn't JSON RPC already an assumption for `Transport`? Is it even meant to work with anything other than JSON RPC?

In any case, I'd include at least RPC in the name.